### PR TITLE
ABC-123 Add handling for loading custom emojis asynchronously

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
This PR does the following:
* Modified emojis to show blank until either the emoji metadata is loaded or found not to exist
* Adds an action when the channel info is shown to load any missing emojis in the channel header
* Adds a server request to emoji autocomplete (similar to how it's done for users)
* Adds custom emoji paging to the emoji picker

Two things to note:
1. I did not add any clearing of the new `nonExistentEmoji` entity when bringing the app to the foreground. The consequence of this is that if someone posts `:some_emoji_that_does_not_exist:`, you switch your app to the background, someone adds that emoji, you're not going to get the new emoji until close and re-open the app. I don't think it's worth clearing it every time you switch to the foreground but let me know if this is unacceptable
2. If someone deletes an emoji while you're not connected to the WS on the app, your store is going to think that emoji still exists and try to load an image for it. This will persist until next log out/in. Not sure how the app handles this for other entities and whether it's acceptable or not

This is my first RN PR, so let me know if I need to be doing something differently :)

#### Ticket Link
https://mattermost.atlassian.net/browse/ABC-123

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes

#### Device Information
This PR was tested on: iPhone 6 simulator, Nexus 6P

### CODE REVIEW

**Comment by jwilander** [comment]
Feel free to give an initial review but I need to add support for old pre 4.7 servers per https://pre-release.mattermost.com/core/pl/cnpsw5qdgifeur8c9haku96raa

**Comment by jwilander** [comment]
Ready for dev review, but waiting on mattermost/mattermost-redux#389

File: app/components/autocomplete/emoji_suggestion/emoji_suggestion.js [hunk 1]
```diff
@@ -63,18 +72,19 @@ export default class EmojiSuggestion extends Component {
...
+        this.matchTerm = match[3];
 
-        const matchTermChanged = matchTerm !== this.state.matchTerm;
-        if (matchTermChanged) {
-            this.setState({
-                matchTerm
-            });
+        const pre47 = !isMinimumServerVersion(Client4.getServerVersion(), 4, 7);
+
+        if (!pre47 && this.matchTerm !== oldMatchTerm && this.matchTerm.length) {
```
Comment (hmhealey) [inline app/components/autocomplete/emoji_suggestion/emoji_suggestion.js]: Is it possible to break out this logic so we have something like
```
if (pre47) {
    // do one thing
} else {
    // do a similar, but slightly different thing
}
```
? It'll probably result in a bit of code duplication, but it'll make this easier to follow since I'm having a tough time figuring this out

File: app/components/autocomplete/emoji_suggestion/emoji_suggestion.js [hunk 2]
```diff
@@ -63,18 +72,19 @@ export default class EmojiSuggestion extends Component {
...
+            this.props.actions.autocompleteCustomEmojis(this.matchTerm);
+            return;
         }
 
-        if (matchTermChanged) {
-            this.handleFuzzySearch(matchTerm, nextProps);
-        } else if (!matchTerm.length) {
+        if (pre47 || this.props.emojis !== nextProps.emojis) {
+            this.handleFuzzySearch(this.matchTerm, nextProps);
+        } else if (!this.matchTerm.length) {
```
Comment (hmhealey) [inline app/components/autocomplete/emoji_suggestion/emoji_suggestion.js]: I can create a separate ticket for this since I think it existed in the old version, but this looks like it's changing the state on every store change under certain circumstances. I think it's also doing it in the `if (!match || this.state.emojiComplete) {` block above

File: app/components/autocomplete/emoji_suggestion/emoji_suggestion.js [hunk 3]
```diff
@@ -63,18 +72,19 @@ export default class EmojiSuggestion extends Component {
...
+
+        if (!pre47 && this.matchTerm !== oldMatchTerm && this.matchTerm.length) {
+            this.props.actions.autocompleteCustomEmojis(this.matchTerm);
+            return;
         }
 
-        if (matchTermChanged) {
-            this.handleFuzzySearch(matchTerm, nextProps);
-        } else if (!matchTerm.length) {
+        if (pre47 || this.props.emojis !== nextProps.emojis) {
```
Comment (enahum) [inline app/components/autocomplete/emoji_suggestion/emoji_suggestion.js]: previously this was `if (matchTerm !== this.state.matchTerm)` why the change on the if statement?

File: app/mattermost.js [hunk 4]
```diff
@@ -90,6 +92,7 @@ export default class Mattermost {
 
         setJSExceptionHandler(this.errorHandler, false);
         setNativeExceptionHandler(this.nativeErrorHandler, false);
+        setSystemEmojis(EmojiIndicesByAlias);
```
Comment (enahum) [inline app/mattermost.js]: this should be done once the store has been rehydrated.

File: app/components/autocomplete/emoji_suggestion/emoji_suggestion.js [hunk 5]
```diff
@@ -63,18 +72,19 @@ export default class EmojiSuggestion extends Component {
...
-        const matchTerm = match[3];
+        const oldMatchTerm = this.matchTerm;
+        this.matchTerm = match[3];
 
-        const matchTermChanged = matchTerm !== this.state.matchTerm;
-        if (matchTermChanged) {
-            this.setState({
-                matchTerm
-            });
+        const pre47 = !isMinimumServerVersion(Client4.getServerVersion(), 4, 7);
```
Comment (hmhealey) [inline app/components/autocomplete/emoji_suggestion/emoji_suggestion.js]: Isn't this stored in the redux store somewhere? It seems weird to access the client directly here.

It also might be nice to have this as a helper function with a clear name to make it clear it's about emojis

File: app/components/emoji/emoji.js [hunk 6]
```diff
@@ -157,6 +145,29 @@ export default class Emoji extends React.PureComponent {
         // force a new image to be rendered when the size changes
         const key = Platform.OS === 'android' ? (height + '-' + width) : null;
 
+        if (!imageUrl) {
+            let BlankComponent;
+            if (Platform.OS === 'android') {
```
Comment (hmhealey) [inline app/components/emoji/emoji.js]: Is this easier if we just use an Image for both platforms? I don't know if using a View on iOS makes much of a difference

File: app/components/emoji/index.js [hunk 7]
```diff
@@ -4,12 +4,40 @@
...
+
+    let imageUrl = '';
+    let isCustomEmoji = false;
+    let displayTextOnly = false;
+    if (EmojiIndicesByAlias.has(emojiName)) {
+        const emoji = Emojis[EmojiIndicesByAlias.get(emojiName)];
+        imageUrl = Client4.getSystemEmojiImageUrl(emoji.filename);
+    } else if (customEmojis.has(emojiName)) {
+        const emoji = customEmojis.get(emojiName);
+        imageUrl = Client4.getCustomEmojiImageUrl(emoji.id);
```
Comment (hmhealey) [inline app/components/emoji/index.js]: Maybe I had been thinking about this differently before, but weren't we adding a route to get a custom emoji's image by name? This looks like we need to load the emoji definition before we can load the emoji image instead of being able to load them in parallel

File: app/components/emoji_picker/emoji_picker.js [hunk 8]
```diff
@@ -317,6 +358,18 @@ export default class EmojiPicker extends PureComponent {
         this.sectionList = c;
     };
 
+    renderFooter = () => {
+        if (!this.state.missingPages) {
+            return null;
+        }
+
+        return (
+            <View style={{flex: 1, alignItems: 'center'}}>
```
Comment (hmhealey) [inline app/components/emoji_picker/emoji_picker.js]: This should go in the StyleSheet

File: app/components/emoji_picker/emoji_picker_row.js [hunk 9]
```diff
@@ -20,7 +21,7 @@ export default class EmojiPickerRow extends Component {
     }
 
     shouldComponentUpdate(nextProps) {
-        return this.props.items.length !== nextProps.items.length;
+        return !shallowEqual(this.props.items, nextProps.items);
```
Comment (hmhealey) [inline app/components/emoji_picker/emoji_picker_row.js]: Does checking this shallowly improve performance by making previous rows rerender less when loading new emojis or is it possibly worse do to shallowly checking constantly

Comment (hmhealey) [reply app/components/emoji_picker/emoji_picker_row.js]: Actually, this is probably better. It's not connected to the store, so this isn't being called anywhere near as often as it would otherwise

File: app/store/index.js [hunk 10]
```diff
@@ -68,7 +68,27 @@ export default function configureAppStore(initialState) {
             };
         },
         null,
-        {whitelist: ['views']} // Only run this filter the views state (or any other entry that ends up being named views)
+        {whitelist: ['views']} // Only run this filter on the views state (or any other entry that ends up being named views)
+    );
+
+    const emojiBlackList = {nonExistentEmoji: true};
```
Comment (hmhealey) [inline app/store/index.js]: One day, we need to find a better library for this. It would be so much nicer if we could just list the parts of the store to be ignored like
```
const ignoredState = {
    'state.entities.emojis.emojiBlackList',
    ...
};
```
instead of having to construct this big function for it

File: app/mattermost.js (hunk 4 above)
Comment (hmhealey) [reply app/mattermost.js]: Does it need to be? This doesn't come from the store, and it's not written there either

Comment (enahum) [reply app/mattermost.js]: What is the redux action doing then?

Comment (jwilander) [reply app/mattermost.js]: It's not a true action, it's not dispatching anything just setting the system emojis to a global (not in the store since they're more of a constant)

File: app/components/autocomplete/emoji_suggestion/emoji_suggestion.js (hunk 3 above)
Comment (jwilander) [reply app/components/autocomplete/emoji_suggestion/emoji_suggestion.js]: For the pre 4.7 case I think you're right I need to check that here

File: app/components/emoji/index.js (hunk 7 above)
Comment (jwilander) [reply app/components/emoji/index.js]: That route didn't get added. We can chat about it some more if you want but I think trying to do both in parallel might be unneeded complexity. Maybe we can do it without pulling the emoji metadata at all but there's some nuances there that I think we'd have to work through

File: app/components/emoji_picker/emoji_picker.js (hunk 8 above)
Comment (jwilander) [reply app/components/emoji_picker/emoji_picker.js]: I copied that from another component doing it this way but I'll throw it into the stylesheet if that's the proper way

File: app/store/index.js (hunk 10 above)
Comment (jwilander) [reply app/store/index.js]: Agreed, it's super annoying to do it this way

File: app/components/emoji/emoji.js (hunk 6 above)
Comment (jwilander) [reply app/components/emoji/emoji.js]: Felt weird to me use an Image component when there was no image to display, let me try it with an Image component on iOS and if that works fine then I'll just do that instead

File: app/components/autocomplete/emoji_suggestion/emoji_suggestion.js (hunk 5 above)
Comment (jwilander) [reply app/components/autocomplete/emoji_suggestion/emoji_suggestion.js]: It is in the redux store, I'll pass it in as a prop

**Comment by jwilander** [comment]
@hmhealey @enahum updates made

**Comment by hmhealey** [comment]
Marking as Ready to Merge in case Joram wants to makes any changes based on Elias's comments

**Comment by jwilander** [comment]
Haha yeah I can do those two little things real quick

File: app/components/autocomplete/emoji_suggestion/emoji_suggestion.js [hunk 11]
```diff
@@ -63,18 +72,33 @@ export default class EmojiSuggestion extends Component {
...
+        // If we're server version 4.7 or higher
+        if (isMinimumServerVersion(this.props.serverVersion, 4, 7)) {
+            if (this.matchTerm !== oldMatchTerm && this.matchTerm.length) {
+                this.props.actions.autocompleteCustomEmojis(this.matchTerm);
+                return;
+            }
+
+            if (this.props.emojis !== nextProps.emojis) {
+                this.handleFuzzySearch(this.matchTerm, nextProps);
+            } else if (!this.matchTerm.length) {
```
[the remark this answers is not in the document]
Comment (jwilander) [reply app/components/autocomplete/emoji_suggestion/emoji_suggestion.js]: Prefer to keep it separate since their are slight differences and it'll make it cleaner to remove it once we stop support pre-4.7 servers
